### PR TITLE
Application priority bug

### PR DESCRIPTION
**Pull request by Snorre98**
:bulb:  replaced this:https://github.com/Samfundet/Samfundet4/pull/1594

There was a bug which made applicant priority of positions incorrect. This is because when priority is manipulated the total count of applications was including withdrawn applications.

This was the case both when a new application was created, but also when applications where withdrawn. 

There also is a bug where  the priority is set incorrectly in the scenario when the applicant applies for a position which they previously had withdrawn the application from (could also be described as "re-activating" the application). The application gets the priority it had when it was withdrawn, which seems like unwanted behavior. I think the behavior should be that the "re-activated" application should get the lowest priority, like when the applicant applies for a position for the first time.

### CODE REVIEW

File: backend/samfundet/models/recruitment.py [hunk 1]
```diff
@@ -430,18 +434,26 @@ def save(self, *args: tuple, **kwargs: dict) -> None:  # noqa: C901
...
+                user=self.user, recruitment=self.recruitment, withdrawn=False
+            ).count()
             # Set the applicant_priority to the number of applications + 1 (for the current application)
-            self.applicant_priority = current_applications_count + 1
+            self.applicant_priority = current_non_withdrawn_applications_count + 1
         # If the application is saved without an interview, try to find an interview from a shared position.
         if self.withdrawn:
+            # when an application is witdrawn the organize_priorities_on_withdrawal signal is called, this must happen post save
+            # this if-statement makes sure that the recruiter_status/priority is "resets" on withdrawal
+            # TODO: DO WE WANT TO SET IT TO NOT WANTED AND AUTOMATIC REJECTION? WOULD NOT_SET AND NOT_SET BE BETTER?
```
Comment (Snorre98) [inline backend/samfundet/models/recruitment.py:452]: TODO: do we actually set NOT_WANTED and AUTOMATIC_REJECTION here?

File: backend/samfundet/models/recruitment.py [hunk 2]
```diff
@@ -472,6 +484,7 @@ def get_total_applications(self) -> int:
         return RecruitmentApplication.objects.filter(user=self.user, recruitment=self.recruitment, withdrawn=False).count()
 
     def update_applicant_state(self) -> None:
+        # TODO: DO WE WANT TO CONSIDER WITHDRAWN APPLICATIONS HERE:
```
Comment (Snorre98) [inline backend/samfundet/models/recruitment.py:487]: TODO: should we add `withdraw=False` on the filters on line 488 and 490?

File: backend/samfundet/models/tests/test_recruitment.py [hunk 3]
```diff
@@ -1093,3 +1093,119 @@ def test_position_must_have_single_owner(fixture_recruitment_position: Recruitme
...
+    # Re-fetch active applications.
+    active_apps = RecruitmentApplication.objects.filter(user=fixture_user, recruitment=fixture_recruitment, withdrawn=False)
+    active_count = active_apps.count()
+    max_priority = max(app.applicant_priority for app in active_apps)
+    # Now there should only be one active application and its applicant_priority should be 1.
+    assert active_count == 1
+    assert max_priority == 1, f'After withdrawal, expected highest priority to be 1, but got {max_priority}'
+
+
+def test_reapplication_updates_priority(
```
Comment (Snorre98) [inline backend/samfundet/models/tests/test_recruitment.py:1166]: This test fails because the priority is not set correctly when the applicant re-activates an application. In such a case the applicant would have withdrawn, but applied for the same position again.

File: backend/samfundet/signals.py [hunk 4]
```diff
@@ -85,3 +85,30 @@ def application_applicant_rejected_or_accepted(sender: RecruitmentApplication, i
...
+        except RecruitmentApplication.DoesNotExist:
+            # If the record does not exist in the database, set old_instance to None.
+            # This might occur if the record was deleted or not yet committed.
+            old_instance = None
+        # If we have an existing record, and it was previously withdrawn,
+        # but now the updated instance is not withdrawn, then the application is being reactivated.
+        if old_instance and old_instance.withdrawn and not instance.withdrawn:
+            # Call a helper method to set the reactivated application's priority
+            # (e.g., by assigning it the lowest active priority plus one).
+            RecruitmentApplication._set_reactivated_priority(instance)
```
Comment (Snorre98) [inline backend/samfundet/signals.py:114]: Im not sure I like this solution

File: backend/samfundet/models/recruitment.py [hunk 5]
```diff
@@ -430,18 +434,26 @@ def save(self, *args: tuple, **kwargs: dict) -> None:  # noqa: C901
...
-            self.applicant_priority = current_applications_count + 1
+            self.applicant_priority = current_non_withdrawn_applications_count + 1
         # If the application is saved without an interview, try to find an interview from a shared position.
         if self.withdrawn:
+            # when an application is witdrawn the organize_priorities_on_withdrawal signal is called, this must happen post save
+            # this if-statement makes sure that the recruiter_status/priority is "resets" on withdrawal
+            # TODO: DO WE WANT TO SET IT TO NOT WANTED AND AUTOMATIC REJECTION? WOULD NOT_SET AND NOT_SET BE BETTER?
             self.recruiter_priority = RecruitmentPriorityChoices.NOT_WANTED
+
             self.recruiter_status = RecruitmentStatusChoices.AUTOMATIC_REJECTION
```
Comment (Snorre98) [inline backend/samfundet/models/recruitment.py:455]: TODO: do we actually want to set NOT_WANTED and AUTOMATIC_REJECTION here?

File: backend/samfundet/models/recruitment.py [hunk 6]
```diff
@@ -471,7 +483,19 @@ def get_total_interviews(self) -> int:
...
+        # Determine the current highest (lowest ranking) priority.
+        max_priority = active_apps.order_by('-applicant_priority').first().applicant_priority if active_apps.exists() else 0
+        # Set the reactivated application's priority to one more than the current maximum.
+        instance.applicant_priority = max_priority + 1
+
     def update_applicant_state(self) -> None:
+        # TODO: DO WE WANT TO CONSIDER WITHDRAWN APPLICATIONS HERE:
         applications = RecruitmentApplication.objects.filter(user=self.user, recruitment=self.recruitment).order_by('applicant_priority')
         # Get top priority
         top_wanted = applications.filter(recruiter_priority=RecruitmentPriorityChoices.WANTED).order_by('applicant_priority').first()
```
Comment (Snorre98) [inline backend/samfundet/models/recruitment.py:501]: TODO: should we add withdraw=False on the filters on line 488 and 490?

File: backend/samfundet/models/recruitment.py [hunk 7]
```diff
@@ -358,12 +358,14 @@ def resolve_gang(self, *, return_id: bool = False) -> Gang | int:
...
+            'applicant_priority'
+        )
+        for i in range(len(non_withdrawn_applications_for_user)):
             correct_position = i + 1
-            if applications_for_user[i].applicant_priority != correct_position:
-                applications_for_user[i].applicant_priority = correct_position
-                applications_for_user[i].save()
+            if non_withdrawn_applications_for_user[i].applicant_priority != correct_position:
+                non_withdrawn_applications_for_user[i].applicant_priority = correct_position
+                non_withdrawn_applications_for_user[i].save()
```
Comment (emsoraffa) [inline backend/samfundet/models/recruitment.py:368]: Kunne kanskje vært gjort litt mer lesbart. også bulk oppdatere for å minimere queries, noe sånn som:

for i, app in enumerate(applications, 1):
            app.applicant_priority = i

RecruitmentApplication.objects.bulk_update(applications, ['applicant_priority'])

Comment (Snorre98) [reply backend/samfundet/models/recruitment.py:368]: Jeg vurdere å lage et issue for å refaktorere RecruitmentPosition etter dette. Koden er litt forvirrende, så den hadde hatt godt av litt hjelpemetoder. 

Prøver å unngå å gjøre store endringer i denne PR-en, slik at det blir enklere å review med tanke på bugen jeg prøver å løse.

**Review by nemisis84 — CHANGES_REQUESTED** [review]
Haven't taken a deep look into this., but leaving a comment anyway The algorithm itself is probably fine and it seems to work in the browser. 

The test should be covering and documenting the edge cases. However, I think all tests should have a docstring describing the test case.

**Review by nemisis84 — APPROVED** [review]
Only approving because I have to. But nice that you added docstrings

**Review by robines — APPROVED** [review]
